apps/onnx/model.cpp

```diff
@@ -0,0 +1,549 @@
+    }
+}
+
+static Halide::Type onnx_type_to_halide_type(int t) {
+    switch (t) {
+    case onnx::TensorProto::FLOAT:
+        return Halide::Type(halide_type_float, 8 * sizeof(float), 1);
+    case onnx::TensorProto::UINT8:
+        return Halide::Type(halide_type_uint, 8 * sizeof(uint8_t), 1);
+    case onnx::TensorProto::INT8:
+        return Halide::Type(halide_type_int, 8 * sizeof(int8_t), 1);
+    case onnx::TensorProto::UINT16:
+        return Halide::Type(halide_type_uint, 8 * sizeof(uint16_t), 1);
+    case onnx::TensorProto::INT16:
+        return Halide::Type(halide_type_int, 8 * sizeof(int16_t), 1);
+    case onnx::TensorProto::INT32:
+        return Halide::Type(halide_type_int, 8 * sizeof(int32_t), 1);
+    case onnx::TensorProto::INT64:
+        return Halide::Type(halide_type_int, 8 * sizeof(int64_t), 1);
+    case onnx::TensorProto::BOOL:
+        return Halide::Type(halide_type_uint, 1, 1);
+    case onnx::TensorProto::DOUBLE:
+        return Halide::Type(halide_type_float, 8 * sizeof(double), 1);
+    case onnx::TensorProto::UINT32:
+        return Halide::Type(halide_type_uint, 8 * sizeof(uint32_t), 1);
+    case onnx::TensorProto::UINT64:
+        return Halide::Type(halide_type_uint, 8 * sizeof(uint64_t), 1);
+    default:
+        throw std::domain_error("Unsupported output type");
+    }
+}
+
+static std::vector<py::array> run(
+    HalideModel pipeline,
+    const std::vector<py::array> &inputs, const std::string &device) {
+    if (inputs.size() == pipeline.model->inputs.size()) {
+        for (int i = 0; i < inputs.size(); ++i) {
+            const std::string &input_name = pipeline.input_names[i];
+            prepare_py_array_input(pipeline, inputs[i], input_name);
+        }
+    } else {
+        throw std::invalid_argument(
+            "Expected " + std::to_string(pipeline.model->inputs.size()) +
+            " numpy arrays but got " + std::to_string(inputs.size()));
+    }
+
+    // Return a list of numpy.ndarray (one per external output)
+    const int num_outputs = pipeline.output_names.size();
+
+    std::vector<Halide::Buffer<>> outputs(num_outputs);
+    for (int i = 0; i < num_outputs; ++i) {
+        Tensor node = pipeline.model->outputs.at(pipeline.output_names.at(i));
+        std::vector<int> output_shape;
+        const int rank = node.shape.size();
+        for (int j = 0; j < rank; ++j) {
+            const int64_t *dim = Halide::Internal::as_const_int(node.shape[j]);
+            if (!dim) {
+                throw std::invalid_argument(
+                    "Couldn't statically infer dim " + std::to_string(j) +
+                    " of output " + std::to_string(i));
+            }
+            output_shape.push_back(*dim);
+        }
+        outputs[i] = Halide::Buffer<>(
+            onnx_type_to_halide_type(pipeline.output_types[i]), output_shape);
+    }
+    Halide::Realization real(outputs);
+    Halide::Target tgt;
+    tgt.set_feature(Halide::Target::NoAsserts, true);
```

Comment:
Using NoAsserts is often a Bad Idea -- if you really need to add it (and have done the performance measurement to verify that is actually worthwhile), it's essential that you document why you know it to be important here (ie how much performance gain it provides), and also how you know that it is safe to do so (ie the pipeline in question is provably safe in all cases).

Reply:
We actually changed that recently and we have NoAsserts for NDEBUG mode.

Reply:
Wait what? I strenuously object to this. Strenuously. This should be discussed widely.

Reply:
CC:@benoitsteiner IIRC jiting a pipeline with assertions was taking long time so we want to compile fast in NDEBUG mode ?!

Reply:
tying Halide::NoAsserts to NDEBUG is not the right answer. If you need to use Halide::NoAsserts, that's a decision that should be made indepdendently from NDEBUG.

Reply:
I think using NoAsserts at all was a hold-over from when we were having llvm problems related to the number of assertions. We should remove this now that we've fixed that.